### PR TITLE
Loop invariant detection.

### PR DESCRIPTION
**Pull request by clyben**
This PR adds loop_invariant.egg which detects the invariant parts inside a Theta node. 
An Operand into a Theta node is invariant if it goes through the Theta, where the i-th output is (Arg i).
The program creates a top-level demand on a Theta node and propagates the demand down to every subpart of the node until it meets the base case (constants or (Arg i)) and records it. Then the program propagates up along with the previously made demands (on Body/Expr/Operand) while recording if that Body/Expr/Operand is constituted sorely by invariant parts.

### CODE REVIEW

File: src/rvsdg/egglog_optimizer/loop_invariant.egg [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+(relation find_inv_body (Body Body))
+;;				body1: context  body2: who owns the vec
+;; IDK if the owner necessary. If two VecOperand in the same region
+;; owned by different body, do it need to be considered seperately?
+(relation find_inv_vo (Body Body VecOperand))
+
+(relation is_inv_expr (Body Expr))
+(relation is_inv_oprd (Body Operand))
+(relation is_inv_body (Body Body))
+(relation is_inv_vo (Body Body VecOperand))
```
Comment (yihozhang) [inline src/rvsdg/egglog_optimizer/loop_invariant.egg]: The second Body may not be necessary.

File: src/rvsdg/egglog_optimizer/loop_invariant.egg [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+(relation arg_inv_helper (VecOperand Body i64))
```
Comment (yihozhang) [inline src/rvsdg/egglog_optimizer/loop_invariant.egg:19]: Describe what it(s every argument) does.

File: src/rvsdg/egglog_optimizer/loop_invariant.egg [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+(rule ((= theta (Theta pred inputs outputs))) 
+	((arg_inv_helper outputs theta 0)))
+
+(rule ((arg_inv_helper (VO outputs) theta i)
+		(< i (vec-length outputs))
```
Comment (yihozhang) [inline src/rvsdg/egglog_optimizer/loop_invariant.egg]: mixed use of space and tab

File: src/rvsdg/egglog_optimizer/loop_invariant.egg [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+		(= (vec-get outputs i) (Arg i)))
+	((arg_inv theta i)))
+
+(rule ((arg_inv_helper (VO outputs) theta i)
+		(< i (vec-length outputs))) 
+	((arg_inv_helper (VO outputs) theta (+ i 1))))
+
+;; top level entry ;
+(rule ((= theta (Theta pred inputs output))
+       (arg_inv theta i))
```
Comment (yihozhang) [inline src/rvsdg/egglog_optimizer/loop_invariant.egg]: Leave a comment for why this is useful for efficiency

File: src/rvsdg/egglog_optimizer/loop_invariant.egg [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+
+;; OperandGroup
+(rule ((find_inv_body theta body)
+		(= body (OperandGroup vo)))
+	((find_inv_vo theta body vo)))
+
+;; vo -> op
+(relation find_inv_vo_helper (VecOperand Body Body i64))
+
+(rule ((find_inv_vo	theta owner vo)) ((find_inv_vo_helper theta owner vo 0)))
```
Comment (yihozhang) [inline src/rvsdg/egglog_optimizer/loop_invariant.egg:61]: tab

File: src/rvsdg/egglog_optimizer/loop_invariant.egg [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+		(= body (Gamma pred inputs outputs))) 
+	((find_inv_oprd theta pred) (find_inv_vo theta body inputs)))
+
+;; OperandGroup
+(rule ((find_inv_body theta body)
+		(= body (OperandGroup vo)))
+	((find_inv_vo theta body vo)))
+
+;; vo -> op
+(relation find_inv_vo_helper (VecOperand Body Body i64))
```
Comment (yihozhang) [inline src/rvsdg/egglog_optimizer/loop_invariant.egg:59]: clarify what each column does

**Comment by clyben** [comment]
I need to modify is_inv_body/expr/oprd from relation to function to 0/1 since in code motion I will also need to find if expr/body/.. are not invariant.

File: src/rvsdg/egglog_optimizer/loop_invariant.egg [hunk 7]
```diff
@@ -0,0 +1,168 @@
...
+(function is_inv_oprd (Body Operand) i64 :unextractable :merge (max old new))
+(function is_inv_body (Body Body) i64 :unextractable :merge (max old new))
+(function is_inv_vo (Body VecOperand) i64 :unextractable :merge (max old new))
+ 
+(relation arg_inv (Body i64))
+;; VecOperand: output of Theta node.
+;; If the node link directly to input, the input operand is invariant.
+;; Body : context
+;; i64: iterator index.
+(relation arg_inv_helper (VecOperand Body i64))
```
Comment (oflatt) [inline src/rvsdg/egglog_optimizer/loop_invariant.egg:20]: Put a TODO here- I think can replace this with @rtjoa's idea of having a table with a range of integers

File: src/rvsdg/egglog_optimizer/loop_invariant.egg [hunk 8]
```diff
@@ -0,0 +1,168 @@
...
+        (= oprd (Project i inv)))
+    ((set (is_inv_oprd body oprd) 1)) :ruleset loop_inv_detect)
+
+;; OperandGroup
+(rule ((find_inv_body theta body)
+        (= body (OperandGroup vo))
+        (= 1 (is_inv_vo theta vo)))
+    ((set (is_inv_body theta body) 1)) :ruleset loop_inv_detect)
+
+;; a theta is invariant in a context when all input are inv.
```
Comment (oflatt) [inline src/rvsdg/egglog_optimizer/loop_invariant.egg:154]: Is this rule just unsound then? If it is, we should not include it.

**Review by oflatt — CHANGES_REQUESTED** [review]
My main concern with this PR is that it doesn't include any tests- could you add small test, one negative and one positive case at least?

**Comment by clyben** [comment]
@oflatt I think I accidentally removed the test in merging....

File: src/rvsdg/egglog_optimizer/loop_invariant.egg [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+;; Body: context, Operand: the invariant parts, i64: the port number.
+(relation theta_invariant_port (Body Operand i64))
+
+;(rule ((arg_inv theta i)) ((theta_inv_port theta (Arg i) i)))
+
+;; can merge the theta not in the program but in somewhere else?
+;(function equiv_theta (Body Body) :unextractable)
+
+;; doing motions
+;; an operand is moved when it's on the edge of invariant and variants
```
Comment (oflatt) [inline src/rvsdg/egglog_optimizer/loop_invariant.egg]: delete these

File: src/rvsdg/egglog_optimizer/loop_invariant.egg [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+(rule ((is_inv_vo_helper body (VO vec) i)
+        (= true (is_inv_operand body operand))
+        (= operand (vec-get vec i)))
+    ((is_inv_vo_helper body (VO vec) (+ i 1))) :ruleset fast-analyses)
+
+(rule ((is_inv_vo_helper body (VO vec) i)
+        (= i (vec-length vec)))
+    ((set (is_inv_vo body (VO vec)) true)) :ruleset fast-analyses)
+
+;; match on every binary op, see loop_invariants.rs
```
Comment (oflatt) [inline src/rvsdg/egglog_optimizer/loop_invariant.egg:125]: another commented out rule here

File: src/rvsdg/egglog_optimizer/loop_invariant.egg [hunk 11]
```diff
@@ -0,0 +1,274 @@
...
+        (= operand (Node body))) 
+    ((find_inv_body theta body)) :ruleset fast-analyses)
+
+;; TODO in some case it might be the body is not invariant but ith output is.
+(rule ((find_inv_operand theta operand)
+        (= operand (Project i body))) 
+    ((find_inv_body theta body)) :ruleset fast-analyses)
+
+;; vo -> op
+(relation is_inv_vo_helper (Body VecOperand i64))
```
Comment (oflatt) [inline src/rvsdg/egglog_optimizer/loop_invariant.egg:112]: need a comment- this proves everything up to an index is invariant?

File: src/rvsdg/egglog_optimizer/loop_invariant.egg [hunk 12]
```diff
@@ -0,0 +1,274 @@
...
+(relation arg_inv_helper (VecOperand Body i64))
+
+;; in default, when there is a find, set is_inv to false
+(rule ((find_inv_expr body term)) ((set (is_inv_expr body term) false)) :ruleset fast-analyses)
+(rule ((find_inv_operand body term)) ((set (is_inv_operand body term) false)) :ruleset fast-analyses)
+(rule ((find_inv_body body term)) ((set (is_inv_body body term) false)) :ruleset fast-analyses)
+(rule ((find_inv_vo body term)) ((set (is_inv_vo body term) false)) :ruleset fast-analyses)
+
+
+(rule ((= theta (Theta pred inputs outputs)))
```
Comment (oflatt) [inline src/rvsdg/egglog_optimizer/loop_invariant.egg:28]: Change this to also have a clause `(VecOperand-get outputs index)`

File: src/rvsdg/tests.rs [hunk 13]
```diff
@@ -1241,3 +1241,150 @@ fn rvsdg_body_contains_operand_group() {
         .parse_and_run_program(EGGLOG_OPERAND_GROUP_PROGRAM)
         .unwrap();
 }
+
+#[test]
+fn rvsdg_loop_inv_detect_simple() {
+    const EGGLOG_THETA_PROGRAM1: &str = r#"
+    (let t1 (Theta 
+            (Node (PureOp (beq (BoolT) (Node (PureOp (bdiv (IntT) (Arg 2) 
+                                                                  (Node (PureOp (bmul (IntT) (Node (PureOp (bsub (IntT) (Arg 1)
```
Comment (oflatt) [inline src/rvsdg/tests.rs:1250]: bad indentation

**Review by oflatt — CHANGES_REQUESTED** [review]
Small changes requested

**Review by oflatt — APPROVED** [review]
Great work!
